core/common/src/internal/format/parser/ParserOperation.kt

```diff
@@ -139,6 +139,130 @@ internal class UnconditionalModification<Output>(
     }
 }
 
+internal class TimeZoneParserOperation<Output>(
+    private val setter: AssignableField<Output, String>
+) : ParserOperation<Output> {
+
+    override fun consume(storage: Output, input: CharSequence, startIndex: Int): ParseResult {
+        if (startIndex >= input.length) return ParseResult.Error(startIndex) { "Unexpected end of input" }
+        val lastMatch = validateTimezone(input, startIndex)
+        if (lastMatch > startIndex) {
+            setter.setWithoutReassigning(storage, input.substring(startIndex, lastMatch), startIndex, lastMatch)
+            return ParseResult.Ok(lastMatch)
+        }
+        return ParseResult.Error(startIndex) { "Invalid timezone format" }
+    }
+
+    companion object {
+        enum class State {
+            START,
+            AFTER_PREFIX,
+            AFTER_SIGN,
+            AFTER_HOUR,
+            AFTER_COLON,
+            END,
+            INVALID
+        }
+
+        private const val INVALID_TIME = -1
+
+        private fun validateTimezone(input: CharSequence, startIndex: Int): Int {
+            var index = startIndex
+            var lastValidIndex = startIndex
+            var hours = INVALID_TIME
+
+            fun hasEnoughChars(length: Int) = index + length <= input.length
+
+            fun validatePrefix(length: Int, validValues: List<String>): Boolean {
+                if (hasEnoughChars(length) && input.substring(index, index + length) in validValues) {
+                    index += length
+                    lastValidIndex = index
+                    return true
+                }
+                return false
+            }
+
+            fun validateTimeComponent(
+                length: Int, maxValue: Int, additionalValidation: (Int) -> Boolean = { true }
+            ): Int {
+                if (!hasEnoughChars(length) || input.slice(index..<(index + length)).any { !it.isDigit() }) {
+                    return INVALID_TIME
+                }
+                val value = input.substring(index, index + length).toInt()
+                if (value > maxValue || !additionalValidation(value)) return INVALID_TIME
+                index += length
+                lastValidIndex = index
+                return value
+            }
+
+            fun validateHH() = validateTimeComponent(2, 18)
+            fun validateH() = validateTimeComponent(1, 9)
+            fun validateMM() = validateTimeComponent(2, 59) { minutes -> hours < 18 || minutes == 0 }
```

Comment:
There are also seconds (`UTC+011530`, `UTC+01:15:30` can both be passed to `TimeZone.of`), but before implementing their support, let's have a discussion tomorrow about which behaviors we want.